### ISSUE TITLE
Tags system

### ISSUE THREAD

**Issue by mk4111** [emfoundation/ce100-app#50]
A new issue has been created to discuss the data model for the two-level tags system.

We need:
- [ ] to provide a unique ID to each of the parent/child tags
- [ ] to relate the parent tags with the corresponding child tags
- [ ] a way in which we can tell which of the parent/child tags are enabled(/disabled).

We hope that the examples below illustrate the options that we have come up with to meet the above requirements are as follows:

#### Option 1: 
1 hash for parent-tags, 1 hash for child-tags, 1 set for enabled parent-tags, n sets for enabled child-tags
```
parentTags: {
  0: 'Energy',
  1: 'Environment,
  2: 'Education'
}

parentTagsKeyNumber-childTag: {   
  0: 'Oil',
  1: 'Gas',
  2: 'Coal',
  3: 'Biofuel'
}  
//  e.g. "0-childTag" would point towards parentTags[0], which is 'Energy' 

parentTags-enabled: [ parentKey1, parentKey2 ] 
// e.g. "parentTags-enabled: [ 0, 1 ]" would identify the parentTags "Energy" and "Environment" as enabled.

parentTagsKeyNumber-enabled: [ childKey1, childKey2 ]
// e.g. "0-enabled: [ 0, 3 ]" would identify the childTags 'Oil' and 'Biofuel' of the parentTag "Energy" as enabled.
```

The issue with this is that there will be lots of sets, as there will be one for every parent-tag...
Sorry for the convoluted explanation, but I hope that made sense...

#### Option 2: 
We were considering nesting an object specifying the name of the parent-tag, relevant child-tags and the enabled/disabled options in one big hash like that below, but thought it would be annoying to have to iterate and JSON.parse the nested object to figure out whether to display the tags or not.

e.g. 
tags: {
  0: JSON.stringify(object)
}

where object is:

{
  Parent: "Energy",
  Children: {
    0: { name: "Oil", enabled: 1 }
    1: { name: "Gas", enabled: 0 }
  }
}

If you could help us figure out what the best option for this, using Redis, would be, we'd be most grateful! Thank you!!

**Comment by asimonw** [comment]
@mk4111 @nelsonic My feeling (but I'm no Redis expert by any means) is that option 1 is a bit clumsy. One way I've seen this parent/child thing done before is by storing all tags in a flat array, every item being a hash with a link to its parent

```
{
    name: string,
    enabled: bool,
    parent: parent_id,
}
```

Items without a parent (i.e. the parent tags themselves) simply have a null value (or -1 if you want to have a pure type). 

This makes it performant to associate an id with a display name in the UI. For example, let's say if you have a list of tag ids for a challenge, you just need to get the object with the right id from the array, which I assume is O(1). Building up the view "Add tag (children)" in Zeplin might be less straight forward, but by using `Array.prototype.filter` judiciously, it's not that hard. 

The only issue with this might be if we would like to be able to re-order tags (by drag and drop, an upcoming feature we're considering), but I think that should also still be ok. Adding tags or assigning tags to a different parent is really easy with this structure.

But very keen for @nelsonic to shoot this down if there's a technical issue with it.

**Comment by nelsonic** [comment]
@asimonw my suggestion is to "MVP" the tagging by storing the entire tag tree in a .json file the way it was in the previous project: https://ce100.ellenmacarthurfoundation.org/json/topics.json

**Comment by asimonw** [comment]
@nelsonic At least, that structure works really well in Handlebars (and JS in general of course). Is there a problem with storing it like this in Redis or are you concerned about the time it will take to build the admin interface? It could be fine to not have an admin UI for it for the MVP. It won't change that often and updating a .json file is easy enough. 

To Marie's point about deep copying (and carefully tracking change events), that's pretty much unavoidable anyway, right?

**Comment by mk4111** [comment]
Closed as #60 was agreed on.
